libs/parAlmond/src/agmgSetup/formAggregates.cpp

```diff
@@ -294,4 +324,258 @@ void formAggregates(parCSR *A, parCSR *C,
   delete C;
 }
 
-} //namespace parAlmond
\ No newline at end of file
+/*****************************************************************************/
+// Alan's "locally partial strong connected nodes" (LPSCN) algorithm.
+
+typedef struct{
+	dlong  index;
+	dlong  Nnbs;
+	dlong LNN;
+} nbs_t;
+
+
+int compareNBSmaxLPSCN(const void *a, const void *b)
+{
+	nbs_t *pa = (nbs_t *)a;
+	nbs_t *pb = (nbs_t *)b;
+
+	if (pa->Nnbs + pa->LNN < pb->Nnbs + pb->LNN)	return +1;
+	if (pa->Nnbs + pa->LNN > pb->Nnbs + pb->LNN)	return -1;
+
+	if (pa->index < pa->index )	return +1;
+	if (pa->index > pa->index )	return -1;
+
+	return 0;
+}
+
+
+int compareNBSminLPSCN(const void *a, const void *b)
+{
+	nbs_t *pa = (nbs_t *)a;
+	nbs_t *pb = (nbs_t *)b;
+
+	if (pa->Nnbs + pa->LNN < pb->Nnbs + pb->LNN)	return -1;
+	if (pa->Nnbs + pa->LNN > pb->Nnbs + pb->LNN)	return +1;
+
+	if (pa->index < pa->index )	return +1;
+	if (pa->index > pa->index )	return -1;
+
+	return 0;
+}
+
+static void formAggregatesLPSCN(parCSR *A, parCSR *C,
+                         hlong* FineToCoarse,
+                         hlong* globalAggStarts,
+                         setupAide options)
+{
+  int rank, size;
+  MPI_Comm_rank(A->comm, &rank);
+  MPI_Comm_size(A->comm, &size);
+
+  const dlong N   = C->Nrows;
+  const dlong M   = C->Ncols;
+  const dlong diagNNZ = C->diag->nnz;
+
+  //  printf("\n-----------------------------------------\n");
+  //printf("rank = %d (LQHSSN) =>  N = %d \t M = %d    ",rank,N,M);
+  //printf("\n-----------------------------------------\n");
+
+  dlong   *states = (dlong *)    calloc(M, sizeof(dlong));    //  M > N
+
+  for(dlong i=0; i<N; i++)   // initialize states to -1
+    states[i] = -1;
+
+  hlong *globalRowStarts = A->globalRowStarts;
+
+  // construct the local neigbors
+  nbs_t *V = (nbs_t *) calloc(N,sizeof(nbs_t));
+
+  for(dlong i=0; i<N; i++){
+	V[i].index    = i;
+	V[i].Nnbs     = C->diag->rowStarts[i+1] - C->diag->rowStarts[i];
+	dlong dist     = C->diag->cols[C->diag->rowStarts[i+1]-1] - C->diag->cols[C->diag->rowStarts[i]];
+	if (dist  > 0 )
+	  V[i].LNN   =  V[i].Nnbs/dist;
+	else
+	  V[i].LNN = 0;
+  }
+
+
+  int MySort = 1;
+  //options.getArgs("SORT",MySort);
+
+  // sort the fake nbs
+  if (options.compareArgs("PARALMOND LPSCN ORDERING", "MAX")) {
+    if (MySort>0)		qsort(V,N,sizeof(nbs_t),compareNBSmaxLPSCN);
+  } else if (options.compareArgs("PARALMOND LPSCN ORDERING", "MIN")) {
+    if (MySort>0)		qsort(V,N,sizeof(nbs_t),compareNBSminLPSCN);
+  } else {
+    // TODO:  Calling exit() within a library is not a good way to handle errors...
+    printf("ERROR:  Bad value for option PARALMOND LPSCN ORDERING.\n");
+    exit(-1);
+  }
+
+  dlong R_num = 0; 	// number of non-isolated nodes to be used
+
+  for (dlong i=0;i<N;i++)
+	if(V[i].Nnbs>1)		R_num++;
+
+
+  dlong R_nodes[N];   // N is number of local elements so at most N aggregates
```

Comment:
Currently fixing this, but for future note: since N can be large this should not be allocated on the stack. Use malloc